lib/TypeNames/ModelToPTMLTypeHelpers.cpp

```diff
@@ -49,38 +51,40 @@ TypeInlineHelper::TypeInlineHelper(const model::Binary &Model) {
 const GraphInfo &TypeInlineHelper::getTypeGraph() const {
   return TypeGraph;
 }
-const TypeSet &TypeInlineHelper::getTypesToInline() const {
+const DefinitionSet &TypeInlineHelper::getTypesToInline() const {
   return TypesToInline;
 }
 const TypeToNumOfRefsMap &TypeInlineHelper::getTypeToNumOfRefs() const {
   return TypeToNumOfRefs;
 }
 
 /// Collect candidates for emitting inline types.
-TypeSet TypeInlineHelper::findTypesToInline(const model::Binary &Model,
-                                            const GraphInfo &TypeGraph) {
-  std::unordered_map<const model::Type *, uint64_t> Candidates;
-  std::set<const model::Type *> ShouldIgnore;
+DefinitionSet TypeInlineHelper::findTypesToInline(const model::Binary &Model,
+                                                  const GraphInfo &TypeGraph) {
+  std::map<const model::TypeDefinition *, uint64_t> Candidates;
+  std::set<const model::TypeDefinition *> ShouldIgnore;
 
   // We may find a struct that represents stack type that is being used exactly
   // once somewhere else in Types:, but we do not want to inline it if that is
   // the case.
-  for (auto &Function : Model.Functions()) {
-    if (not Function.StackFrameType().empty()) {
-      const model::Type *StackT = Function.StackFrameType().getConst();
-      ShouldIgnore.insert(StackT);
-    }
-  }
-
-  for (const UpcastablePointer<model::Type> &T : Model.Types()) {
-    for (const model::QualifiedType &QT : T->edges()) {
-      auto *DependantType = QT.UnqualifiedType().get();
-      if (llvm::isa<model::RawFunctionType>(T.get())
-          or llvm::isa<model::CABIFunctionType>(T.get())
-          or llvm::isa<model::TypedefType>(T.get())) {
+  for (auto &Function : Model.Functions())
+    if (not Function.StackFrameType().empty())
+      ShouldIgnore.insert(Function.stackFrameType());
+
+  for (const model::UpcastableTypeDefinition &T : Model.TypeDefinitions()) {
+    for (const model::Type *Edge : T->edges()) {
+      const model::TypeDefinition *DependantType = Edge->skipToDefinition();
+      if (DependantType == nullptr)
+        continue; // Skip types without definitions, like primitives.
+
+      if (llvm::isa<model::RawFunctionDefinition>(T.get())
+          or llvm::isa<model::CABIFunctionDefinition>(T.get())
+          or llvm::isa<model::TypedefDefinition>(T.get())) {
         // Used as typename.
         ShouldIgnore.insert(DependantType);
-      } else if (isCandidateForInline(DependantType)) {
+      } else if (llvm::isa<model::StructDefinition>(DependantType)
+                 or llvm::isa<model::UnionDefinition>(DependantType)
+                 or llvm::isa<model::EnumDefinition>(DependantType)) {
```

Comment:
In the end, I got rid of `canBeInlined` and replaced all of its usages (including this one) with `declarationIsDefinition` since we no longer have primitives differentiating the two.